go/vt/wrangler/tablet.go

```diff
@@ -201,3 +208,23 @@ func (wr *Wrangler) VReplicationExec(ctx context.Context, tabletAlias *topodatap
 	}
 	return wr.tmc.VReplicationExec(ctx, ti.Tablet, query)
 }
+
+func (wr *Wrangler) isMasterTablet(ctx context.Context, ti *topo.TabletInfo) (bool, error) {
+	// Tablet record claims to be non-master, we believe it
+	if ti.Type != topodatapb.TabletType_MASTER {
+		return false, nil
+	}
+	si, err := wr.ts.GetShard(ctx, ti.Keyspace, ti.Shard)
+	if err != nil {
+		// strictly speaking it isn't correct to return false here, the tablet status is unknown
+		return false, err
+	}
+	// Tablet record claims to be master, and shard record matches
+	if topoproto.TabletAliasEqual(si.MasterAlias, ti.Tablet.Alias) {
+		return true, nil
+	}
+	// Shard record has another tablet as master, so check MasterTermStartTime
+	// If tablet record's MasterTermStartTime is equal to or later than the one in the shard record, then tablet is master
+	// !Before == Equal || After
+	return !logutil.ProtoToTime(ti.MasterTermStartTime).Before(logutil.ProtoToTime(si.MasterTermStartTime)), nil
```

Comment:
Elsewhere (like in shard_sync.go) we only "steal" mastership if the timestamp is `After()`, not if it's merely equal. The idea was, in the case of a tie, whoever updates the shard record first wins. Otherwise the shard record could oscillate as two tablets fight each other.

Also a nit: can we break out the `ProtoToTime` calls into vars to make the final expression easier to read?

Reply:
That is correct, but it would be wrong to only test for `After()` here. When a tablet takes over as master, we set `MasterTermStartTime` on the shard record to be the same as the tablet's `MasterTermStartTime`.
https://github.com/vitessio/vitess/blob/master/go/vt/vttablet/tabletmanager/shard_sync.go#L171

Reply:
Thinking about it further, I think I understand what you are getting at. Since shard.MasterAlias != tablet.Alias, this tablet can be the true master only if it's MasterTermStartTime > shard's MasterTimeStartTime. Is that correct?